README
			This is the README file for 


                      BBBB   U   U  GGGGG  SSSSS  X   X 
                      B   B  U   U  G      S       X X  
                      BBBB   U   U  G GGG  SSSSS    X  
                      B   B  U   U  G   G      S   X X  
                      BBBB    UUU   GGGGG  SSSSS  X   X 



1) Author & Copyright

	bugsx was written by 

		(C) Copyright 1995-1997 Robert Gasch (Robert_Gasch@peoplesoft.com)
		http://www.peoplesoft.com/peoplepages/g/robert_gasch/index.htm

		(C) Copyright 1990-1997 Joshua R. Smith (jrs@media.mit.edu) 
		http://physics.www.media.mit.edu/~jrs

	is copyright by the author(s). I (Robert_Gasch@peoplesoft.com) maintain 
	the X11 port and Joshua maintains any other ports. See section 4 
	and 17 for more info.

2) Distribution Policy

 	Permission to use, copy, modify and distribute this software for
	any purpose and without fee is hereby granted, provided that this
	copyright notice appear in all copies as well as supporting 
	documentation. All work developed as a consequence of the use of 
	this program should duly acknowledge such use.

	See the GNU General Public Licence for more information.

	What does this mean?
	This means that you can pretty much do with it what you want as 
	long as you don't charge substantially more than the distribution 
	costs for it (of course you have to make the source available). 

	This makes software distributions like Walnut Creek or Infomagic 
	perfectly acceptable. 

	It does NOT matter wether you use it at home or in a comercial 
	environment. 

	If you want to include it in a major commercial distribution you 
	need my permission to do so.

3) Guarantees

	bugsx should work all right *but*

	No representations are made about the suitability of this software 
	for any purpose. This software is provided "as is" without express 
	or implied warranty.

4) What is bugsx?

	bugsx runs under MIT's X11 window system. It was written under 
	UNIX but should be easily portable.  It is a program which draws 
	the biomorphs based on parametric plots of Fourier sine and cosine
        series and let's you play with them using the genetic algorithm.

	The original version which ran under Suntools and XViews was 
	written by Joshua R. Smith sometime 1990. See the 'credits' 
	section for more details. 

	For more information about the theoretic backgrounds of 
	bugsx consult Joshua R. Smith's paper distributed with this 
	program in gzip'ed postscript format as bugs.ps.gz. You have 
	to uncompress this file with 'gunzip' before you can print it. 
	Gunzip should be available at an archive near you.

	The paper can also be gotten from:
	http://physics.www.media.mit.edu/~jrs/biomorphs.ps

	I (Robert_Gasch@peoplesoft.com) maintain the X11 port and 
	Joshua (jrs@media.mit.edu) maintains any other ports. 
	See section 4, and 17 for more info. 

5) Version Number

	The current version number is 1.08

6) Where can I get Bugsx from?

	Bugsx can be gotten from the following FTP sites:
	media.mit.edu 	(/physics/dist/bugs)
	ftp.x.org	(/contrib/applications)

7) What if you change the program?

	please mail me (Robert_Gasch@peoplesoft.com) the changes(!) as I'd like to 
	act as the central source for bugsx. This will also enable me to 
	include your changes in a future release. If your changes are 
	generic, I'll pass the changes on so that other versions of BUGS 
	can also benefit from your input.

8) Systems currently supported

	Any resonable UNIX-type system running X11 should be able to 
	compile and run bugsx. I've tested it on the following 
	systems:

	Linux
	AIX 

9) How do I get started?

	This section assumes that you have all the necessary files. (The 
	file "Manifest" contains a list of distributed files.)

	a) Customize the Imakefile to suit your system setup. On most systems 
	this should work OK as it is.

	b) Make the Makefile by typing 
		xmkmf
	If your system does not support generating makefiles from 
	Imakefiles use "Makefile.std" as a template to build your
	own. 

	c) If your X headers do not define APP_DEFAULTS_DIR and your default 
	X Resource file is not 
		/usr/X11R6/lib/X11/app-defaults/
	edit bugsx.h and modify the definition for APP_DEFAULTS_DIR. On 
	most systems things should be OK the way they are, but this gives 
	you the possibility to customize this.

	d) Once you have a valid Makefile (see step a&b) build bugsx by typing
		make

	Once you have created the executable, you're ready to go. The
	executable will be created in the directory in which the source
	for bugsx resides. 

10) Brief users guide

	You can get a list of supported options by typing one of the
	following:
		bugsx -?
		bugsx -help
		bugsx help

	Once you start bugsx, you get the main window containing 
	the menu options at the top and the evenly distributed drawing 
	windows below the menus. 

	The Menus are:

		Initialize - Initializes the gene pool 
		Breed 	   - Breed the selected organisms.  
		Quit       - exits bugsx

	Once you have chosen "Initialize" and the first set of drawings 
	is shown, you can select the ones you wish to be included in 
	the cross-breeding by clicking (with the left button) on a 
	drawing window. The selected organisms will be drawn inverse. 

	Clicking on Breed will cross breed the selected organisms using 
	the genetic algorithm. If you have not selected at least two 
	organisms, a set of at least two will be randomly chosen and 
	used for the breeding. 

	The following keys can be used to control bugsx:
		q - quit bugsx 
		i - initialize population
		b - breed

	Interesting/useful sample settings:
		bugsx -interval 3 -number 4 -batch -nobreed -segments 500
		bugsx -interval 3 -batch -showbreed 
		bugsx -seed 10 


10) Command line parameters

	bugsx accepts the following parameters as command line options:

	+rv 			reverse video (use to override xrdb entry)
	+synchronous		syncronous mode (use to override xrdb entry)
	-? 			help
	-background <arg>	backgound color
	-batch 			run program in batch mode
	-bg <arg> 		same as -background
	-bordercolor <arg>	border color
	-borderwidth <arg> 	border width
	-cycle <arg>		re-initialize population after n batch turns
	-display 		display
	-extend_print		show extended reproduction info while running
	-fg <arg>		same as -forground
	-font <arg>		font
	-foreground <arg>	forground color (also file system bar color)
	-geometry <arg>		geometry 
	-help 			help
	-iconic			iconic
	-interval <arg>		interval used per turn
	-mb			show menu border
	-name <arg>		run bugsx under this name
	-nobreed 		do not breed when running in batch mode
	-number <arg>		number of biomorphs to draw (must be a square #)
	-printpop		print the population when breeding
	-rv 			reverse video
	-seed <arg>		use this seed for random number generator
	-segments <arg>		use this many segments to draw an organism
	-showbreed		show breeding subpopulation when in batch mode
	-showgenes		show a graphic representation of the genes
	-synchronous		synchronous mode
	-v 			verbose
	-xrm			make no entry in resrouce database
	help 			help


11) XResources 

        bugsx recognizes the following XResources. Usually bugsx will search 
	for resources under the program name but you can override this with 
	the -name flag. If you do not wish to use a specific application 
	defaults file, you can execute xrdb -merge to merge your resource 
	specifications into the XResource database. 

	bugsx first checks in the directory pointed at by the environment 
	variable XAPPLRESDIR. If this doesn't yield any resource definitions 
	it checks the APP_DEFAULTS_DIR. This is defined in your headers or 
	in bugsx.h. If you want to change this you'll have to recompile bugsx.

	background		universal backgound color
	batch			run program in batch mode
	batchbreed		do not breed when running in bactch mode
	borderColor		border color
	borderWidth		border width
	cycle			re-initialize population after n batch turns
	display			display
	extend_print		show extended reproduction info while running
	font			font
	foreground		universal foreground color 
	help			show help
	mainWin.geometry	main window geometry
	iconic			start program in iconic mode 
	interval		interval used per turn
	minimize		minimize window size 
	menuborder		draw menu borders
	name			run bugsxunder this name
	number			number of biomorphs to draw (must be a square #)
	printpop		print the population when breeding
	reverseVideo		reverse video
	seed			use this seed for random number generator
	segments		use this many segments to draw an organism
	showbreed		show breeding subpopulation when in batch mode
	showgenes		show a graphic representation of the genes
	synchronous		syncronous mode
	verbose			verbose mode 
	
12) Troubleshooting

	Problem: Compile failes with a problem with the catlist() function
		definition.
	Solution: In terms of portability this is the most fragile part 
		of bugsx. I've done my best to make this portable, but 
		there may be systems whose compiler simply throws up once 
		he sees catlist. If you do have problems with catlist(), 
		there are a number of things you can try:
		1) Try using another compiler (gcc generally accepts the 
		catlist() function as-is). 
		2) Undef/Define  __STDC__ in your compiler directives. Most 
		compilers do this via the -U flag. 
		If these fail, you'll have to adapt catlist() for your 
		machine. In that case, accept my apologies for the problems
		and please send me a mail telling me what you did so that I 
		can integrate your fix into the sources, sparing others of 
		this misery.

	Problem: Bugsx won't find my resource definitions.
	Solution: Bugsx handles resources as follows:
		a) bugsx checks in the directory pointed at by the environment 
		   variable XAPPLRESDIR. If this doesn't yield any resource 
		   definitions we go to step b. If bugsx ignores your personal 
		   resrource definitions, make sure you have XAPPLRESDIR 
		   set to the proper value and exorted. 
	        b) bugsx checks the APP_DEFAULTS_DIR. This is defined in your 
		   headers or in bugsx.h. If you want to change this you'll 
		   have to recompile bugsx.
		The reasoning behind this approach is to allow user preferences 
		to take preference over the system preferences. 

	Problem: Bugsx seems to start OK but I don't get the window.
	Solution: Check if your DISPLAY variable is set correctly (you
		can always use the -v falg to see if it does start up 
		correctly).

13) I've found a bug. What now?

	Read section 11 (Troubeshooting) to see if it falls under one of
	the problems listed there. If not either fix it yourself or mail
	me. If my time allows it, I'll try to fix things.  


debian/changelog
bugsx (1.08-13) experimental; urgency=low

  * Moved from monolithic patch to topgit+quilt.
  * Moved to source format 3.0 (quilt).
  * Started using dh.
  * Pointed debian/copyright at GPL-1, not GPL.
  * Mentioned non-freeness in debian/copyright.
  * Bumped Standards-Version to 3.9.1.

 -- Marcin Owsiany <porridge@debian.org>  Mon, 22 Nov 2010 18:21:12 -0800

bugsx (1.08-12) unstable; urgency=medium

  * Added a dependency on libxext-dev, which is needed now that libx11-dev
    does not depend on it (closes: 475065)
  * Bumped Standards-Version to 3.7.3 (no changes needed)
  * Capitalized "needs" value and added a longtitle to the menu file
  * Changed clean target to run distclean when Makefile exists, instead of
    ignoring its exit status
  * Fix "dependency" spelling errors in changelog

 -- Marcin Owsiany <porridge@debian.org>  Wed,  9 Apr 2008 19:53:34 +0100

bugsx (1.08-11) unstable; urgency=low

  * Added a desktop file (closes: 432462) and a call to dh_desktop
  * Switched from deprecated dh_movefiles to dh_install
  * Added a menu icon, just a simple rescaled snapshot of a part of the
    screen, with transparent background.

 -- Marcin Owsiany <porridge@debian.org>  Fri, 10 Aug 2007 16:19:48 +0100

bugsx (1.08-10) unstable; urgency=low

  * Patched xbutil.c to avoid a buffer overflow (closes: 431374)
  * Fixed the menu file (outdated path, missing quotes). Also changed section
    to Games/Toys, since it's not really a math application as such
  * Bumped Standards-Version to 3.7.2 (no changes needed)
  * Fixed section in manpage to 1x to match filename
  * Updated DH compat version and build-dep to 5
  * Updated year in the copyright
  * Added a uscan watch file (closes: 377346)

 -- Marcin Owsiany <porridge@debian.org>  Mon,  2 Jul 2007 10:03:06 +0100

bugsx (1.08-9) unstable; urgency=high

  * Transition to xorg 7 (RC, thus high urgency):
    - debian/bugsx.files: changed binary and manpage path to match new
      standard-FHS X layout
    - build-depend on xutils-dev for imake (closes: 364041, 348163)
  * A proper review of the license text resulted in:
    - added the whole text of "distribution policy" and my comment to
      debian/copyright
    - moved to non-free
    - more details in thread starting at:
      http://lists.debian.org/debian-legal/2006/04/msg00221.html
  * Improved description synopsis a little bit

 -- Marcin Owsiany <porridge@debian.org>  Fri, 28 Apr 2006 20:27:24 +0100

bugsx (1.08-8) unstable; urgency=low

  * Change build-dependency on xlibs-dev to libx11-dev, since the former was
    removed from the archive. (closes: 346628)

 -- Marcin Owsiany <porridge@debian.org>  Sat, 14 Jan 2006 14:08:50 +0100

bugsx (1.08-7) unstable; urgency=low

  * New maintainer. (closes: #86636)
  * Now using debhelper:
     - corrected update-menu invocation
     - don't create/remove the /usr/doc symlink
  * Pay attention to DEB_BUILD_OPTIONS
  * Upgraded standards-version
  * Moved the binary and manpage to /usr/X11R6/{bin,man} (this is
    where the imake-generated makefile would put them) and corrected
    the path in menu file

 -- Marcin Owsiany <porridge@debian.org>  Sun, 13 Apr 2003 16:31:18 +0200

bugsx (1.08-6) unstable; urgency=low

  * Orphaned this package.

 -- Adrian Bunk <bunk@fs.tum.de>  Fri, 25 Jan 2002 14:05:00 +0100

bugsx (1.08-5) unstable; urgency=low

  * New maintainer. (closes: #68303)
  * Remove the Makefile in the clean target.
  * Added "-isp" to dpkg-gencontrol in debian/rules.
  * Strip bugsx better.
  * /usr/doc -> /usr/share/doc
  * /usr/man -> /usr/share/man
  * Added build dependencies.
  * Standards-Version: 3.1.1

 -- Adrian Bunk <bunk@fs.tum.de>  Sat,  3 Feb 2001 19:47:08 +0100

bugsx (1.08-4) unstable; urgency=low

  * Maintainer set to Debian QA Group.
  * Removed README.debian.
  * Removed the emacs config directives from debian/changelog.

 -- Martin Michlmayr <tbm@cyrius.com>  Fri, 22 Dec 2000 15:51:08 +0100

bugsx (1.08-3) unstable; urgency=low

  * link against glibc 2.1.1
  * add to menu-system

 -- John Lapeyre <lapeyre@physics.arizona.edu>  Thu, 25 Mar 1999 00:26:44 -0700

bugsx (1.08-2) unstable; urgency=low

  * Fixed description.
  * fixed hardcoded egcc.

 -- John Lapeyre <lapeyre@physics.arizona.edu>  Tue, 15 Sep 1998 18:59:44 -0700

bugsx (1.08-1) unstable; urgency=low

  * Initial Release.

 -- John Lapeyre <lapeyre@physics.arizona.edu>  Tue, 25 Aug 1998 12:13:45 -0700



bugsx.c
/* ************************************************************************* *
   bugsx - (C) Copyright 1990-1997 Joshua R. Smith (jrs@media.mit.edu)
	   http://physics.www.media.mit.edu/~jrs

	   (C) Copyright 1995-1997  Robert Gasch (Robert_Gasch@peoplesoft.com)
	   http://www.peoplesoft.com/peoplepages/g/robert_gasch/index.htm

   Permission to use, copy, modify and distribute this software for any 
   purpose and without fee is hereby granted, provided that this copyright
   notice appear in all copies as well as supporting documentation. All
   work developed as a consequence of the use of this program should duly
   acknowledge such use.

   No representations are made about the suitability of this software for
   any purpose. This software is provided "as is" without express or implied 
   warranty.

   See the GNU General Public Licence for more information.
 * ************************************************************************* */




/* *************************************************************************
   This program displays a list of bar graphs for the file systems of the 
   machine you are on. The following switches are supported:

	+rv			reverse video (use to override xrdb entry) 
	+synchronous		syncronous mode (use to override xrdb entry)
	-?			help
	-background <arg>	backgound color 
	-batch			run program in batch mode
	-bg <arg>		same as -background 
	-bordercolor <arg>	border color
	-borderwidth <arg>	border width
	-cycle			# of iterations after which we re-initialize
	-display		display
	-extend_print		show extended reproduction info while running
	-fg <arg>		same as -forground
	-font <arg>		font
	-foreground <arg>	forground color (also file system bar color)
	-geometry <arg>		geometry (will override extreme window sizes to 
					apply reasonable settings)
	-help			help
	-iconic			iconic
	-interval <arg>		interval used per turn
	-mb			show menu border
	-name <arg>		run bugsx under this name
	-nobreed		do not breed when running in batch mode
	-number	<arg>		number of biomorphs to draw (must be a square #)
	-printpop		print the population when breeding
	-rv 			reverse video
	-seed <arg>		use this seed 
	-segments <arg>		use this many segments to draw an organism
	-showbreed		show brreding subpopulation when in batch mode
	-showgenes		show a graphical representation of the genes
	-synchronous		synchronous mode
	-v			verbose
	-xrm			make no entry in resrouce database
	help			help



   Version History:
	10-05-90: 1.00
			First version by Joshua R. Smith, running only 
			under Xview and Suntools.
	21-04-95: 1.01
			First version running under XWindows. 
	24-04-95: 1.02 
			Lots of improvements. Added resource handling,
			-printpop, -extend_print, -number, -batch, 
			-interval, -segments, -nobreed flags.
			Cleaned up some other code.
	26-04-95: 1.03 
			Added -seed, -showbreed flags, lots of parameter 
			sanity checking added.
	26-04-95: 1.04
			Made sure breeding subpop is selected when user 
			clicks on "Initialize" when running in batch mode.
			Fixed bug with handling of -nobreed.
	02-05-95: 1.05
			Added resize support, changed #ifdef __STDC__ to 
			#if __STDC__.
	02-06-95: 1.06
			Changed the #if __STDC__ back to #ifdef __STDC__
			since some compilers rely on the ambigous construct,
			changed HP check to #ifdef __hpux, updated resource 
			handling, updated README file. 
	29-03-96: 1.07
			Changed method for checking if -number is a square 
			number, added some (double) casts for sqrt(),
			got genes to print, added -showgenes, changed 
			address to Robert_Gasch@peoplesoft.com.
	20-08-97: 1.08
			Added code so that you can select organisms to 
			keep prior to re-initializing the gene pool, 
			added -cycle flag. 

  ************************************************************************* */


#include "bugsx.h"


Display			*mydisplay;
char*			myname;		/* Name we hunt resources with */
XrmDatabase		resourceDB;		/* X resources. */
WinType			main_win,		/* main window */
			menu[MENU_ITEMS],	/* menu items */
			draw_win[MAX_DRAW_WINS];/* windows we draw in */
XGCValues		gray_gc_val;		/* GC values for gray fill */
XEvent			myevent;		/* event */
unsigned long		fg, bg;			/* forgrund, background */
unsigned long		warn1col,		/* warn color level 1 */
			warn2col,		/* warn color level 2 */
			warn3col;		/* warn color level 3 */
XSizeHints		mainwin_hint,		/* main Window size hints */
			szhint;			/* detail Window size hints */ 
Pixmap			warn_pix;		/* Warning Pixmap */
Population		G_Population[MAX_POP];	/* Array of Organisms */
Population		G_Kids_Pop[MAX_POP];	/* Next generation */
double			G_fit_thresh,		/* fitness threshold */
			G_pCross,		/* probability of crossover */
			G_pMutation,		/* probability of mutation */
			G_mutation_std,		/* gauss fn STD used to mutate*/
			G_weight[MAX_CHROM_SIZE],/*weight for each term */
			G_sum_weights;		/* Sum of weights -- yscaling */
int			G_size_pop,		/* # organisms in population */
			G_size_breeding_pop,	/* # organisms reproducing */
			G_generation=NOGOOD,	/* # generations so far */
			G_switch_default,	/* switches start on or off */
			G_show_genes,		/* display gene window	*/
			G_current_width,	/* Current size of pixwin */
			G_current_height,
			G_org_height,		/* Size in pixels of organism */
			G_org_width,
			G_x_scale,		/* These are used for scaling */
			G_y_scale,		/*	curves.	*/
			G_x_trans,
			G_y_trans,
			cycle=0,
			verbose=FALSE,
			batch=FALSE, 
			batch_breed=TRUE,
			show_breed=FALSE,
			upd_interval=UPDINTERVAL,
			segments=ORG_SEGMENTS,
			selected[MAX_DRAW_WINS],
			do_big=FALSE,
			do_print_pop=FALSE,
			extend_print=FALSE,
			Draw_Wins=DRAW_WINS,
			Draw_Rows=DRAW_ROWS,
			Draw_Columns=DRAW_COLUMNS,
			menu_border=FALSE;	/* menu border */
long			seed,
			mstime=MSTIME;



static XrmOptionDescRec opTable[] = {
		/* KEEP THIS SORTED BY THE FIRST ARGUMENT! */
/* the char * should be XPointer but many systems don't support it */
{"+rv",		"*reverseVideo", XrmoptionNoArg,	(char *) "off"},
{"+synchronous","*synchronous",	XrmoptionNoArg,		(char *) "off"},
{"-?",		".help",	XrmoptionNoArg,		(char *) "on"},
{"-background",	"*background",	XrmoptionSepArg,	(char *) NULL},
{"-batch",	".batch",	XrmoptionNoArg,		(char *) "on"},
{"-bg",		"*background",	XrmoptionSepArg,	(char *) NULL},
{"-bordercolor","*borderColor",	XrmoptionSepArg,	(char *) NULL},
{"-borderwidth",".borderWidth",	XrmoptionSepArg,	(char *) NULL},
{"-cycle",	".cycle",	XrmoptionSepArg,	(char *) NULL},
{"-display",	".display",     XrmoptionSepArg,	(char *) NULL},
{"-extend_print",".extend_print",XrmoptionNoArg,	(char *) "on"},
{"-fg",		"*foreground",	XrmoptionSepArg,	(char *) NULL},
{"-font",	"*font",	XrmoptionSepArg,	(char *) NULL},
{"-foreground",	"*foreground",	XrmoptionSepArg,	(char *) NULL},
{"-geometry",	"*mainWin.geometry",XrmoptionSepArg,	(char *) NULL},
{"-help",	".help",	XrmoptionNoArg,		(char *) "on"},
{"-interval",	".interval",	XrmoptionSepArg,	(char *) NULL},
{"-iconic",	".iconic",	XrmoptionNoArg,		(char *) "on"},
{"-mb",		".menuborder",	XrmoptionNoArg,		(char *) "on"},
{"-name",	".name",	XrmoptionSepArg,	(char *) NULL},
{"-nobreed", 	".batchbreed",	XrmoptionNoArg,		(char *) "off"},
{"-number",	".number",	XrmoptionSepArg,	(char *) NULL},
{"-printpop",	".printpop",	XrmoptionNoArg, 	(char *) "on"},
{"-rv",		"*reverseVideo",XrmoptionNoArg,		(char *) "on"},
{"-seed",	".seed",     	XrmoptionSepArg,	(char *) NULL},
{"-segments",	".segments",	XrmoptionSepArg,	(char *) NULL},
{"-showbreed",	".showbreed",	XrmoptionNoArg, 	(char *) "on"},
{"-showgenes",	".showgenes",	XrmoptionNoArg, 	(char *) "on"},
{"-synchronous","*synchronous",	XrmoptionNoArg,		(char *) "on"},
{"-v",		".verbose",	XrmoptionNoArg,		(char *) "on"},
{"-xrm",	NULL,		XrmoptionResArg,	(char *) NULL},
{"help",	".help",	XrmoptionNoArg,		(char *) "on"},
};


void main (argc, argv)
int argc;
char **argv;
{
	int 		i=0;
	XrmDatabase 	commandlineDB=NULL;
	char* 		myDisplayName = NULL;
	XrmString 	str_type;
	XrmValue 	value;

	/* *** Try and work out the name the application will run under *** */
	/* *** This needed for finding the correct resources. *** */
	{
	char** ptr = &argv[1];

	for (; *ptr != NULL; ptr++)
	if (strcmp(*ptr, "-name") == 0 && *(ptr + 1) != NULL) 
		{
		myname = *(ptr + 1);
		break;
		}

	if (*ptr == NULL) 
		{
		char* ptr1 = rindex(argv[0], '/');
		if (ptr1)
			myname = ++ptr1;
		else
			myname = argv[0];
		}
	}

	XrmInitialize();
	XrmParseCommand(&commandlineDB, opTable, 
		sizeof(opTable) / sizeof(opTable[0]), myname, &argc, argv);
	if (XrmGetResource(commandlineDB, 
		catlist(myname, ".name", (char*) NULL), "Xfsm.Name", 
		&str_type, &value))
			myname = value.addr;
	if (XrmGetResource(commandlineDB, 
		catlist(myname, ".display", (char*) NULL), "Xfsm.Display", 
		&str_type, &value))
			myDisplayName = value.addr;

	/* *** connect to X Server and assign screen *** */
	if ((mydisplay = XOpenDisplay(myDisplayName)) == NULL) 
		{
		fprintf (stderr, "Error Opening Display ... Exiting\n");
		exit (1);
		}

	/* *** init the simulation constants *** */
	init();

	/* *** process arguments and the resource database *** */
	process_databases (argc, argv, commandlineDB);

	/* *** Turn on X server synchronization if the user wants it. *** */
	XSynchronize(mydisplay, getBoolResource(resourceDB,
		catlist(myname, ".synchronous", (char*) NULL),
		"Xfsm.Synchronous", False));

	init_all_windows();
	srand48 (seed);

	/* *** Check to see of the user wants the colors reversed. *** */
	if (getBoolResource(resourceDB, 
		catlist(myname, ".reverseVideo", (char*) NULL),
		"Xfsm.ReverseVideo", False)) 
		{
		long temp;
		temp = fg; 
		fg = bg ; 
		bg = temp;
		}

	/* *** create Main Program Window and set up the properties. *** */
	{
	XClassHint clshint;
	XWMHints wmhints;

	mainwin_hint.min_width = MIN_WIN_X;
	mainwin_hint.min_height = MIN_WIN_Y;
	mainwin_hint.width = main_win.width;
	mainwin_hint.height = main_win.height;
	mainwin_hint.x = main_win.x;
	mainwin_hint.y = main_win.y; 
	mainwin_hint.flags = main_win.flags | PMinSize | PSize;

	clshint.res_name  = myname;
	clshint.res_class = CLASS_NAME;

	wmhints.input = True; 
	wmhints.initial_state = (getBoolResource(resourceDB,
		catlist(myname, ".iconic", (char*) NULL), 
		"Xfsm.Iconic", False))
      		? IconicState : NormalState;
	wmhints.window_group = main_win.win;
	wmhints.flags = InputHint | StateHint | WindowGroupHint;

	create_window (DefaultRootWindow(mydisplay), &main_win, fg, bg);
	XSetStandardProperties (mydisplay, main_win.win, main_win.text,
		main_win.text, None, argv, argc, &mainwin_hint);
	XSetClassHint(mydisplay, main_win.win, &clshint);
	XSetWMHints(mydisplay, main_win.win, &wmhints);
	}

	if (do_big)
		Draw_Wins=1;
	/* *** create all the buttons using child windows *** */
 	for (i = 0; i < MENU_ITEMS; ++i) 
		create_window (main_win.win, &menu[i], fg, bg); 
	for (i = 0; i < Draw_Wins; i++) 
		create_window (main_win.win, &draw_win[i], fg, bg);

	/* *** Map window definitions onto screen *** */
	XMapRaised (mydisplay, main_win.win);
	XMapSubwindows (mydisplay, main_win.win);

	randomize_pop();

#ifdef DEBUG
	fprintf (stdout, "Setup is Done!\n");
#endif 

/* ************************************************************ */
/* ************** setup is done - main event loop ************* */
/* ************************************************************ */
	
	do_event_loop (argc, argv);

	/* *** we are done - destroy all the windows *** */

	destroy_menu (menu, MENU_ITEMS);
	destroy_menu (draw_win, Draw_Wins); 
	XFreeGC (mydisplay, main_win.gc);
	XDestroyWindow (mydisplay, main_win.win);
	XCloseDisplay (mydisplay);
	exit (0);
}



/* *** This is the main event loop. Here we handle events and deal *** */
/* ***                 with all the user inputs we get             *** */
void do_event_loop (argc, argv)
int argc;
char **argv;
{
	XEvent	tevent;
	int	iter=0, count, cont, this_item, i, done=FALSE; 
	char	text[10];
	long	target = time(NULL);

    while (!done)	
	{
	if (batch)
		{
		/* *** here we wait while there are no events *** */
        	/* *** and count out down our interval.       *** */
        	while (XPending (mydisplay) == 0)
               		{
			struct timeval  sleept;
			fd_set reads;

			FD_ZERO(&reads);
			FD_SET(ConnectionNumber(mydisplay), &reads);
			sleept.tv_sec = target - time(NULL);
			sleept.tv_usec = 0;
#ifdef __hpux
			select(ConnectionNumber(mydisplay) + 1, (int *) 
				&reads, NULL, NULL, &sleept);
#else
			select(ConnectionNumber(mydisplay) + 1, &reads, 
				NULL, NULL, &sleept);
#endif
			/* *** interval has passed *** */
			if ((time(NULL)) >= target)
				{
				grow_pop();
				target = target + upd_interval;
				if (!batch_breed)
					randomize_pop();
				else
					{
					breed();
					select_breeding_sub_pop();
					}

				if (++iter == cycle)
					{
					printf ("HERE\n");
					randomize_pop();
					iter=0;
					}
				}
			}
		}

	/* *** read the next event *** */
	XNextEvent (mydisplay, &myevent);

	switch (myevent.type) 
	{
	/* *** expose event -> redraw the window *** */
	 case Expose:
	/* ********************************************* *
	 *  since a window can generate multiple expose  *
	 *  events we can collapse these into one redraw *
	 * ********************************************* */
		count=0;
		do 
			{
			if (XPending (mydisplay) == 0)
				break;
			XPeekEvent (mydisplay, &tevent);
			if (tevent.xexpose.window ==  myevent.xexpose.window) 
				{
				XNextEvent (mydisplay, &myevent);
				count++;
				}
			} 
		while(tevent.xexpose.window==myevent.xexpose.window);

#ifdef DEBUG
		if (!count)
			fprintf(stdout, "got expose event in window: %d\n",
			myevent.xexpose.window);
		else
			fprintf(stdout, 
			"Compressed %d expose events into 1 in window: %d\n",
			count, myevent.xexpose.window);
#endif 

		/* ******************************************* *
		 * ** now deal with the actual expose event ** *
		 * ******************************************* */
		if (myevent.xexpose.window == main_win.win &&
		    myevent.xexpose.count == 0) 
			{
			redraw_main_win ();
			break;
			} 
		cont = expose_menu (menu, MENU_ITEMS);
		if (!cont)
			break;
		if (G_generation != NOGOOD)
			cont = expose_win (draw_win, Draw_Wins);

		break;

	/* *** change work window if main window changes *** */
	case ConfigureNotify:
#ifdef DEBUG
		fprintf(stdout, "got configure notify event in window: %d\n",
			myevent.xconfigure.window);
#endif
		/* *** we don't allow changes in the detail window *** */
		if (myevent.xconfigure.window != main_win.win)
			break;

		/* *** size stays the same -> nothing to do *** */
		if (myevent.xconfigure.width == main_win.width &&
		    myevent.xconfigure.height == main_win.height)
			break;

		XClearWindow (mydisplay, main_win.win);
		handle_resize (myevent.xconfigure.width, 
			       myevent.xconfigure.height);
		break;

	/* *** process keyboard mapping changes *** */
	case MappingNotify:
#ifdef DEBUG
	fprintf(stdout, "got mapping notify event in window: %d\n",
		myevent.xexpose.window);
#endif

		XRefreshKeyboardMapping ((XMappingEvent *) &myevent );
		break;

      /* *** drag in work window *** */
	case MotionNotify:
#ifdef DEBUG
	fprintf(stdout, "got motion notify event in window: %d\n",
		myevent.xbutton.window);
#endif
	break;

	/* *** mouse enters a window *** */
	case EnterNotify:
		this_item = highlight_menu (menu, MENU_ITEMS, TRUE);
		break;

	/* *** Mouse Leaves a window *** */
	case LeaveNotify:
		this_item = highlight_menu (menu, MENU_ITEMS, FALSE);
		break;

	/* *** process mouse-button presses *** */
	case ButtonPress:
#ifdef DEBUG
	fprintf(stdout, "button press (%d) in window: %d\n",
		myevent.xbutton.button, myevent.xbutton.window);
#endif

	if (myevent.xbutton.button == 1 || myevent.xbutton.button == 2) 
		{
		this_item = which_button_pressed (menu, MENU_ITEMS);
		if (this_item == INITIALIZE) 
			{
			randomize_pop();
			grow_pop();
			if (batch)
				select_breeding_sub_pop();
			break;
			} 
		else if (this_item == BREED)
			{
			if (batch || Draw_Wins == 1)
				break;
			breed ();
			break;
			}
		else if (this_item == QUIT) 
			{
			done = TRUE; 
			break;
			}
/*
		else if (this_item == PLUS)
			{
			while (!XPending (mydisplay))
				{
				msleep (mstime);
				if (mstime > 5000)
					mstime-=5000;
				if (segments < MAX_SEGMENTS)
					segments++;
				sprintf (menu[SHOW].text, "%8d", segments);
				XDrawImageString (mydisplay, 
					menu[SHOW].win,
					menu[SHOW].gc, 1, 
					LETTER_ASCENT (menu[SHOW]),
					menu[SHOW].text, 
					strlen (menu[SHOW].text));
				}
			mstime=MSTIME;
			break;
			}
		else if (this_item == MIN)
			{
			while (!XPending (mydisplay))
				{
				msleep (mstime);
				if (mstime > 5000)
					mstime-=5000;
				if (segments > MIN_SEGMENTS)
					segments--;
				sprintf (menu[SHOW].text, "%8d", segments);
				XDrawImageString (mydisplay, 
					menu[SHOW].win,
					menu[SHOW].gc, 1, 
					LETTER_ASCENT (menu[SHOW]),
					menu[SHOW].text, 
					strlen (menu[SHOW].text));
				}
			mstime=MSTIME;
			break;
			}
		else if (this_item == OK)
			{
			grow_pop();
			}
*/

		this_item = which_button_pressed (draw_win, Draw_Wins);
		if (this_item == NOGOOD || G_generation == NOGOOD)
			break;
		else
			highlight_org_window (this_item);
		}
	break;

	/* *** process mouse-button release *** */
	case ButtonRelease:
#ifdef DEBUG
		fprintf(stdout, "button release in window: %d\n", 
			myevent.xbutton.window);
#endif
	break;

	/* *** process Resize *** */
	case ResizeRequest:
#ifdef DEBUG
		fprintf (stdout, "got resize event \n");
#endif
	break;

	/* *** process keyboard input *** */
	case KeyPress:
#ifdef DEBUG
		fprintf (stdout, "got keypress event in window: %d\n", 
			myevent.xkey.window);
#endif
		i = XLookupString ((XKeyEvent *)&myevent, text, 10, NULL, 0 );

		if (text[0] == INIT_KEY)
			{
			randomize_pop();
			grow_pop();
			if (batch)
				select_breeding_sub_pop();
			break;
			}
		else 
		if (text[0] == BREED_KEY)
			{
			if (batch || Draw_Wins == 1)
				break;
			breed ();
			break;
			}
		else
		if (text[0] == QUIT_KEY) 
			{
			done = TRUE;
			break;
			}
		break;
	} /* switch (myevent.type) */
    } /* while (done == 0) */
}
